### DISCUSSION TITLE
Make SQLAlchemy >= 2.0.0

### DISCUSSION THREAD

**Discussion by wu-clan** [fastapi-practices/fastapi_best_architecture#3] [Ideas]
Hi, @downdawn

This issue needs to be fixed: [I6GCQ7](https://gitee.com/wu_cl/fastapi_sqlalchemy_mysql/issues/I6GCQ7) to https://github.com/sqlalchemy/sqlalchemy/discussions/9333

I have to drop using SQLAlchemy-2.0.0b1, and upgrade to SQLAlchemy >= 2.0.8 for backwards compatibility.

Since I prefer to use `MappedAsDataclass`, I now have to retrofit MiXin to work properly, but please note that I cannot use any new features that are only available in python > 3.7

**Comment by downdawn** [comment]
Ok. I didn't encounter this problem, and I started with the latest SQLAlchemy.

This seems to have been fixed in SQLAlchemy 2.0.8.

The DataClassBase and Base options you provide make model definition easier.

**Comment by wu-clan** [comment]
I tested SQLAlchecmy 2.0.8, 2.0.9.

This is an issue that won't be fixed, because SQLAlchemy works in its own dataclass model

Create a new database, confirm your SQLAlchemy version, then perform the migration via alembic, or just start the service and you will find the exception

**Comment by downdawn** [comment]
I missed the discussion at [https://github.com/sqlalchemy/sqlalchemy/discussions/9333](https://github.com/sqlalchemy/sqlalchemy/discussions/9333)

I tested it, and it is indeed normal in 2.0.0b1, and there is no backward compatibility.

So what's the new Mixin workaround?

**Comment by wu-clan** [comment]
Not yet. I‘ve been a little busy lately.

What needs to be solved most is the field order problem of SQAlchemy data mode. We only need to ensure that the MiXin parameters are always arranged at the end of the subclass parameters when they are inherited, not the front.

In this way, there will be no extra work.

**Comment by downdawn** [comment]
When the Python interpreter parses the parameter list, it always parses the parameters in the order of the parent class first, and then the child class.

Unless the fields of the parent class are overridden in the subclass, but that would be meaningless.

**Reply by downdawn** [reply]
update_user set init=False can run successfully

```
update_user: Mapped[Optional[int]] = mapped_column(default=None, init=False, comment='Modifier')
```

**Reply by downdawn** [reply]
![image](https://user-images.githubusercontent.com/41266749/231376052-5934906c-6a6f-4225-8901-4c3f16c9f1f6.png)

[https://docs.sqlalchemy.org/en/20/orm/dataclasses.html](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html)

**Reply by wu-clan** [reply]
> [https://docs.sqlalchemy.org/en/20/orm/dataclasses.html](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html)

Wow, I haven’t read its documentation for a long time. I‘ll check it tonight.
